m3/src/reporter_impl.h

```diff
@@ -0,0 +1,146 @@
+
+namespace tally {
+namespace m3 {
+
+class Reporter::Impl : public tally::StatsReporter {
+ public:
+  Impl(const std::string &host, uint16_t port,
+       const std::map<std::string, std::string> &common_tags,
+       uint32_t max_queue_size, uint16_t max_packet_size);
+
+  ~Impl();
+
+  // Ensure the class is non-copyable.
+  Impl(const Impl &) = delete;
+
+  Impl &operator=(const Impl &) = delete;
+
+  // Methods to implement the StatsReporter interface.
+  std::unique_ptr<tally::Capabilities> Capabilities();
+
+  void Flush();
+
+  void ReportCounter(const std::string &name,
+                     const std::map<std::string, std::string> &tags,
+                     int64_t value);
+
+  void ReportGauge(const std::string &name,
+                   const std::map<std::string, std::string> &tags,
+                   double value);
+
+  void ReportTimer(const std::string &name,
+                   const std::map<std::string, std::string> &tags,
+                   std::chrono::nanoseconds value);
+
+  void ReportHistogramValueSamples(
+      const std::string &name, const std::map<std::string, std::string> &tags,
+      uint64_t bucket_id, uint64_t num_buckets, double buckets_lower_bound,
+      double buckets_upper_bound, uint64_t samples);
+
+  void ReportHistogramDurationSamples(
+      const std::string &name, const std::map<std::string, std::string> &tags,
+      uint64_t bucket_id, uint64_t num_buckets,
+      std::chrono::nanoseconds buckets_lower_bound,
+      std::chrono::nanoseconds buckets_upper_bound, uint64_t samples);
+
+ private:
+  // Run implements the logic of the Reporter, pulling metrics of its queue and
+  // emitting them.
+  void Run();
+
+  // Enqueue adds a metric to the Reporter's queue.
+  void Enqueue(thrift::Metric);
+
+  // Process processes a metric so it can be emitted.
+  void Process(thrift::Metric metric,
+               std::shared_ptr<TCalcTransport> calc_transport);
+
+  // Helper methods used when processing metrics.
+  void ReportMetric(const std::string &name,
+                    const std::set<thrift::MetricTag> tags,
+                    thrift::MetricValue value);
+
+  thrift::MetricValue CreateCounter(int64_t value);
+
+  thrift::MetricValue CreateGauge(double value);
+
+  thrift::MetricValue CreateTimer(std::chrono::nanoseconds value);
+
+  std::set<thrift::MetricTag> ConvertTags(
+      const std::map<std::string, std::string> &tags);
+
+  std::string ValueBucketString(double bucket_bound);
+
+  std::string DurationBucketString(std::chrono::nanoseconds bucket_bound);
+
+  std::string BucketID(uint64_t bucket_id, uint64_t num_buckets);
+
+  std::string FormatDuration(std::chrono::nanoseconds duration, int precision);
+
+  static const char HISTOGRAM_BUCKET_NAME[];
+  static const char HISTOGRAM_BUCKET_ID_NAME[];
+
+  const std::set<thrift::MetricTag> common_tags_;
+  const uint32_t max_queue_size_;
+  const uint16_t max_packet_size_;
+
+  std::condition_variable cv_;
```

Comment:
nit: rename to indicate what the cv is used for - `bg_thread_cv_` maybe?

Reply:
Try using `std::promise<void>` if you only need a one time signal. I used that a lot in Jaeger's C++ client.